### PR TITLE
Feature/class strategy

### PR DESCRIPTION
**Pull request by jaapio**
This pr contains the strategy to create class descriptors.
As discussed it also contains 2 iterators to be able to create property and class constants in a strategy.

### CODE REVIEW

File: src/phpDocumentor/Reflection/Php/Factory/Class_.php [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+    }
+
+    /**
+     * @param Node|PropertyIterator|ClassConstantIterator $stmt
+     * @param StrategyContainer $strategies
+     * @param ClassDescriptor $classDescriptor
+     */
+    private function addCreateAndMember($stmt, StrategyContainer $strategies, ClassDescriptor $classDescriptor)
+    {
+        $strategy = $strategies->findMatching($stmt);
```
Comment (mvriel) [inline src/phpDocumentor/Reflection/Php/Factory/Class_.php]: Wouldn't it be better to move line 123 and 124 to its own method and move the addMethod/Property/Constant calls back to 97, 103 and a specific version for Method there? That saves us a switch statement and a form of duplication

File: src/phpDocumentor/Reflection/Php/Factory/Class_.php [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+    /**
+     * Creates an ClassDescriptor out of the given object.
+     * Since an object might contain other objects that need to be converted the $factory is passed so it can be
+     * used to create nested Elements.
+     *
+     * @param ClassNode $object object to convert to an Element
+     * @param StrategyContainer $strategies used to convert nested objects.
+     * @return ClassDescriptor
+     */
+    public function create($object, StrategyContainer $strategies)
```
Comment (mvriel) [inline src/phpDocumentor/Reflection/Php/Factory/Class_.php]: This method is too large, is it possible to divide it into more specific private methods?

File: src/phpDocumentor/Reflection/Php/Factory/PropertyIterator.php [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+     * @var PropertyNode
+     */
+    private $property;
+
+    /** @var int index of the current propertyProperty to use */
+    private $index = 0;
+
+    public function __construct(PropertyNode $propery)
+    {
+        $this->property = $propery;
```
Comment (mvriel) [inline src/phpDocumentor/Reflection/Php/Factory/PropertyIterator.php]: s/propery/property

File: src/phpDocumentor/Reflection/Php/Factory/ClassConstantIterator.php [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+ * This class acts like a combination of a ClassConst and Const_ to be able to create constant descriptors using a normal strategy.
+ */
+final class ClassConstantIterator implements \Iterator
+{
+    /**
+     * @var ClassConst
+     */
+    private $classConstants;
+
+    /** @var int index of the current propertyProperty to use */
```
Comment (mvriel) [inline src/phpDocumentor/Reflection/Php/Factory/ClassConstantIterator.php]: propertyProperty is probably a copy paste error

File: src/phpDocumentor/Reflection/Php/Factory/ClassConstantIterator.php [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+    /**
+     * Initializes the class with source data.
+     *
+     * @param ClassConst $classConst
+     */
+    public function __construct(ClassConst $classConst)
+    {
+        $this->classConstants = $classConst;
+    }
+
```
Comment (mvriel) [inline src/phpDocumentor/Reflection/Php/Factory/ClassConstantIterator.php]: superfluous whiteline

File: src/phpDocumentor/Reflection/Php/Factory/PropertyIterator.php [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+    }
+
+    public function getName()
+    {
+        return $this->property->props[$this->index]->name;
+    }
+
+    public function getDefault()
+    {
+        return $this->property->props[$this->index]->default;
```
Comment (mvriel) [inline src/phpDocumentor/Reflection/Php/Factory/PropertyIterator.php]: Returns `Node\Expr` but a string representation is wanted

File: src/phpDocumentor/Reflection/Php/Factory/ClassConstantIterator.php [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+namespace phpDocumentor\Reflection\Php\Factory;
+
+use PhpParser\Comment\Doc;
+use PhpParser\Node\Stmt\ClassConst;
+
+/**
+ * This class acts like a combination of a ClassConst and Const_ to be able to create constant descriptors using a normal strategy.
+ */
+final class ClassConstantIterator implements \Iterator
```
Comment (mvriel) [inline src/phpDocumentor/Reflection/Php/Factory/ClassConstantIterator.php]: Missing a `getValue` method; constants have methods that should be interpreted using the PrettyPrinter and returned

File: src/phpDocumentor/Reflection/Php/Factory/Class_.php [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+use PhpParser\Node\Stmt\Property as PropertyNode;
+use PhpParser\Comment\Doc;
+use PhpParser\Node\Stmt\TraitUse;
+
+/**
+ * Strategy to create a ClassDescriptor including all sub elements.
+ */
+final class Class_ implements ProjectFactoryStrategy
+{
+
```
Comment (mvriel) [inline src/phpDocumentor/Reflection/Php/Factory/Class_.php]: superfluous whiteline

**Comment by mvriel** [comment]
@jaapio I have concluded my review, I really like the approach you took with those iterators. That is an innovative way to approach the issue that didn't occur to me and it is _neat_!

During my review I found two or three bugs and a few style issues. I did notice as well that some of the DocBlocks are missing a summary or completely absent; can you add those for me please?

**Comment by mvriel** [comment]
Also: the build fails; is this something I should ignore for now?

**Comment by jaapio** [comment]
I will update the strategy for constants in an other pr. For now this class was required to be able to write an integration test.

For places where you mentioned that the prettyprinter should be used, I will add this to the strategies of the specific element. The iterators should only provide de real values.

**Comment by mvriel** [comment]
Can't merge anymore :)

**Comment by jaapio** [comment]
rebased it.

**Pull request merged**